### ISSUE TITLE
[1.18 regression] False negative with variable increment in a walrus inside a tuple

### ISSUE THREAD

**Issue by hamdanal** [python/mypy#20606]
**Bug Report**

mypy sometimes fails to report errors in assignment expressions inside a tuple.

**To Reproduce**

```python
def condition() -> bool: return False

def fn() -> tuple[int, int]:
    i: int | None = 0 if condition() else None
    return (i, (i := i + 1))

print(fn())
```

**Expected Behavior**

Something like
```
lib.py:5: error: Incompatible return value type (got "tuple[int | None, int]", expected "tuple[int, int]")  [return-value]
lib.py:5: error: Unsupported operand types for + ("None" and "int")  [operator]
lib.py:5: note: Left operand is of type "int | None"
```

**Actual Behavior**

```
Success: no issues found in 1 source file
```

**Your Environment**

- Mypy version used: 1.19.1 and master
- Mypy command-line flags: n/a
- Mypy configuration options from `mypy.ini` (and other config files): n/a
- Python version used: 3.14.2

The weird thing is that if I change assignment target to a different variable like `(i, (j := i + 1))` mypy reports the two errors above.
Possibly related to https://github.com/mypyc/mypyc/issues/1179

**Comment by hamdanal** [comment]
Looks like a regression as this did raise an error in version 1.17.1.

Removed the affects mypyc label as even with version 1.17.1 mypyc has the same problem as in the linked issue.

Bisects to commit a856e559756 (#19767)
/cc @ilevkivskyi as the author of the change

**Comment by ilevkivskyi** [comment]
Oh, after looking at this I discovered there is a whole class of problems in presence of walrus. The point is there are certain scenarios where we visit an expression multiple times (usually to get a better/more user-friendly type inference). For example:
```python
i: int | None = 0 if condition() else None
x: tuple[int, int] | None = (i, (i := 1))  # No error now
```
Addressing this properly is tricky. So for now, I am going to fix most obvious cases and leave a TODO explaining a more permanent solution.

**Comment by ilevkivskyi** [comment]
Here is a fix https://github.com/python/mypy/pull/20622
